src/nwb_conversion_tools/tools/spikeinterface/spikeinterface.py

```diff
@@ -1138,6 +1135,83 @@ def add_units(
+    units_table_name: str = "units",
+    unit_table_description: str = "Autogenerated by nwb_conversion_tools.",
+    write_units_table_in_processing_module: bool = False,
+):
+    """
+    Primary method for writing a SortingExtractor object to an NWBFile.
+    Parameters
+    ----------
+    sorting: SpikeInterfaceSorting
+    save_path: PathType
+        Required if an nwbfile is not passed. The location where the NWBFile either exists, or will be written.
+    overwrite: bool
+        If using save_path, whether or not to overwrite the NWBFile if it already exists.
+    nwbfile: NWBFile
+        Required if a save_path is not specified. If passed, this function
+        will fill the relevant fields within the nwbfile. E.g., calling
+        spikeextractors.NwbRecordingExtractor.write_recording(
+            my_recording_extractor, my_nwbfile
+        )
+        will result in the appropriate changes to the my_nwbfile object.
+    skip_features: list of str
+        Each string in this list that matches a spike feature will not be written to the NWBFile.
+    units_table_name : str (optional, defaults to 'units')
+        The name of the units table. If write_as=='units', then units_table_name must also be 'units'.
+    unit_table_description : str (optional)
+        Text description of the units table; it is recommended to include information such as the sorting method,
+        curation steps, etc.
+    write_units_table_in_processing_module: bool (optional, defaults to False)
+        How to save the units table in the nwb file.
+        - True will save it to the processing module to serve as a historical provenance for the official table.
+        - False will save it to the official NWBFile.Units position; recommended only for the final form of the data.
+    """
+    if isinstance(sorting, SortingExtractor):
+        sampling_frequency = sorting.get_sampling_frequency()
+        if sampling_frequency is None:
+            raise ValueError("Writing a SortingExtractor to an NWBFile requires a known sampling frequency!")
+        checked_sorting = OldToNewSorting(oldapi_sorting_extractor=sorting)
+    else:
+        checked_sorting = sorting
+    unit_ids = checked_sorting.get_unit_ids()
+    sampling_frequency = checked_sorting.get_sampling_frequency()
+
+    if write_units_table_in_processing_module:
+        ecephys_mod = get_module(
+            nwbfile=nwbfile,
+            name="ecephys",
+            description="Intermediate data from extracellular electrophysiology recordings, e.g., LFP.",
+        )
+        if units_table_name not in ecephys_mod.data_interfaces:
+            units_table = pynwb.misc.Units(name=units_table_name, description=unit_table_description)
+            ecephys_mod.add(units_table)
+
+        units_table = ecephys_mod[units_table_name]
+    else:
+        if nwbfile.units is None:
+            nwbfile.units = pynwb.misc.Units(name=units_table_name, description=unit_table_description)
```

Comment:
So the logic here is to write the units table to the NWBFile in memory first, then modify the reference to that? 

Also, how would this modular setup of writing handle a case where I try to write (a) waveforms before writing spike times and (b) can it detect mismatches in shapes between the spiketimes for a unit and the shape of the waveform array?

Reply:
Yes, that's correct. That was the logic as it was before (see 'add_units' function) and I think it plays well with the requirement of choosing where to write the units table ( e.g in a processing module vs the standard place for the canonical table).

Concerning your questions:
a) This does not work right now. As mentioned in #450, `add_units`, in its current state does not support re-writing of any sort. After the next PR in this series, it should handle it OK.
b) I don't know what do you mean here, I have not modified this function at all so I don't know where this is done, is this related to row mismatch across multiple writings? Maybe if you point me how this problem was solved in the previous version I can say whether it would work with the new logic.

Reply:
b) probably was not solved in previous version. Just occurred to me with the new append feature... the dimensionality of waveforms for a single unit should be `(number of spike times, number of frames, number of channels)`. If, going forward, you attempt to add a waveform column after writing spike times, the first dimension should match or else it isn't clear which spike time indices you are subsetting.

Reply:
> That was the logic as it was before (see 'add_units' function) 

The logic on main:
https://github.com/catalystneuro/nwb-conversion-tools/blob/main/src/nwb_conversion_tools/tools/spikeinterface/spikeinterface.py#L1199-L1210 adds the units table object to the nwbfile as the very final step (even after writing waveforms).

I think, if you want to split out the waveform step into a separate function altogether, it would make considerably more sense to have it apply functionally to an input `UnitsTable` object, rather than an `nwbfile` where it then takes all those extra arguments to determine 'which' table within the file to act on. This would also simplify the logic of having to create a table in the event it doesn't exist, which seems a very odd way to go about writing waveforms (harkening to point (a) above).

Reply:
Concerning a) I misunderstood you. Yes, this only keeps backwards compatibility with the previous framework of filling the units table and then immediately add the waveforms. To do this, the proper reference written before by add_units) is fetched and the waveforms are added to this table. 

Concerning the second point in the last response.
Yes, maybe that would make sense. I also thought that it would make sense not only for the `add_waforms` table but also for the `add_units one`.

That said, I think that belongs to a separate PR which we can discuss once we finish refactoring the units_table. The current scope of this is just to separate the two functions while giving backwards compatibility for the framework we had before. 

Once we have an `add_units` function that supports multiple writes we can discuss how the cases that you mention fits into that framework, how they will interact with the refactored add_units function and device a PR specifically to cover those cases that you mention with proper testing.

Reply:
> > That was the logic as it was before (see 'add_units' function)
> 
> The logic on main: https://github.com/catalystneuro/nwb-conversion-tools/blob/main/src/nwb_conversion_tools/tools/spikeinterface/spikeinterface.py#L1199-L1210 adds the units table object to the nwbfile as the very final step (even after writing waveforms).
> 
> I think, if you want to split out the waveform step into a separate function altogether, it would make considerably more sense to have it apply functionally to an input `UnitsTable` object, rather than an `nwbfile` where it then takes all those extra arguments to determine 'which' table within the file to act on. This would also simplify the logic of having to create a table in the event it doesn't exist, which seems a very odd way to go about writing waveforms (harkening to point (a) above).

@CodyCBakerPhD 
Maybe the disagreement comes from here? 
The function has to decide where to write (units vs processing) at the end because it never has to overwrite (if the units table or the processing table already exists it exits without doing anything):

https://github.com/catalystneuro/nwb-conversion-tools/blob/177733629f1cea1a885191f588eb37f1b28c9396/src/nwb_conversion_tools/tools/spikeinterface/spikeinterface.py#L1114

As we separate the `add_waveforms` function this assumption **can not* be hold to be true anymore (as previously `add_units` modify the table), so the logic of fetching the place where the table is to be written needs to be added at the beginning as well.

If I understand you well, you strongly believe that this should not be done inside function but maybe I don't understand you well.

Reply:
OK, here's the clearest plan I can layout for the API-level setup with an eye towards the future additions.

For API:

A WaveformExtractor is essentially a container for a RecordingExtractor and SortingExtractor. Though I can also see cases for it containing direct waveform snippets instead of a recording in cases where the original format didn't keep raw data but did record pre-extracted snippets.

As such, we will want a `write_waveforms(...)` function that similar to `write_recording` and `write_sorting` acts on a file level and is mostly a router to a helper function. That helper function *could* be called `add_waveforms(...)` and it *could* include the kind of argument input setup you have now; my point is however, the design of *that* `add_waveforms` function I'm talking about right now should do the following:

```python
def add_waveforms(...):
    # write recording metadata
    # write the sorting extractor contents from the one contained in the WaveformExtractor, most importantly the spike times
    # then add the waveforms for each of the spike times on that table
```

Now, I am proposing that the logic of the commented blocks would be easiest to implement as follows:

```python
units_table = add_units(...)
add_waveforms_to_units_table(units_table=units_table)
```

Basically, the logic as you have it now, would add the waveform column to a units table independently of the spike times, which opens the door to a long series of complications due to shape mismatch assertions between the two that are already handled by the code being moved. There's that complication on top of all of  the excess logic and input arguments for creating a brand new units table in the event there is no existing table, which absolutely should not happen here.

What I am suggesting, then, as clearly as I can state it, is for you to change what you are currently calling `add_waveforms` to a private function `_add_waveforms_to_table(units_table)` that is still called during `add_units(, write_waveforms=True)` for the case when the input sorting object is an old SortingExtractor that can have unit spike features. Then in a future PR we can add a proper `add_waveforms` that truly behaves as expected according to the outline above.

Reply:
I get you now, I did not get what you are saying in the last paragraph from your previous suggestion. 

Having that private method called from inside the function does seem like a better solution to me that is even less of a change.

Reply:
All right, added.